packages/flutter/lib/src/cupertino/checkbox.dart

```diff
@@ -201,19 +312,72 @@ class _CupertinoCheckboxState extends State<CupertinoCheckbox> with TickerProvid
   @override
   bool? get value => widget.value;
 
-  void onFocusChange(bool value) {
-    if (focused != value) {
-      focused = value;
-    }
+  WidgetStateProperty<Color> get _widgetFillColor {
+    return WidgetStateProperty.resolveWith((Set<WidgetState> states) {
+      if (states.contains(WidgetState.disabled)) {
+        return widget.inactiveColor ?? CupertinoColors.white;
+      }
+      if (states.contains(WidgetState.selected)) {
+        return widget.activeColor ?? CupertinoColors.activeBlue;
+      }
+      return CupertinoColors.white;
+    });
+  }
+
+  WidgetStateProperty<Color> get _widgetCheckColor {
+    return WidgetStateProperty.resolveWith((Set<WidgetState> states) {
+      if (states.contains(WidgetState.disabled) && states.contains(WidgetState.selected)) {
+        return widget.checkColor ?? CupertinoColors.black;
+     }
+      if (states.contains(WidgetState.selected)) {
+        return widget.checkColor ?? CupertinoColors.white;
+      }
+      return CupertinoColors.white;
+    });
   }
 
+  WidgetStateProperty<BorderSide> get _widgetSide {
+    const Color borderColor = CupertinoColors.systemGrey4;
+    return WidgetStateProperty.resolveWith((Set<WidgetState> states) {
+      if (states.contains(WidgetState.disabled)){
+        return const BorderSide(color: borderColor);
+      }
+      if (states.contains(WidgetState.selected) || states.contains(WidgetState.focused)) {
+        return const BorderSide(width: 0.0, color: _kTransparentColor);
+      }
+      return const BorderSide(color: borderColor);
+    });
+  }
+
+    BorderSide? _resolveSide(BorderSide? side, Set<WidgetState> states) {
+      if (side is WidgetStateBorderSide) {
+        return WidgetStateProperty.resolveAs<BorderSide?>(side, states);
+      }
+      if (!states.contains(WidgetState.selected)) {
+        return side;
+      }
+      return null;
+    }
+
   @override
   Widget build(BuildContext context) {
-    final Color effectiveActiveColor = widget.activeColor
-      ?? CupertinoColors.activeBlue;
-    final Color? inactiveColor = widget.inactiveColor;
-    final Color effectiveInactiveColor = inactiveColor
-      ?? CupertinoColors.inactiveGray;
+    // Colors need to be resolved in selected and non selected states separately.
+    final Set<WidgetState> activeStates = states..add(WidgetState.selected);
+    final Set<WidgetState> inactiveStates = states..remove(WidgetState.selected);
```

Comment:
Instead of passing both inactive and active colors to _painter, would it be any cleaner to pass it the WidgetStateProperty and the `states` and let it resolve the color? Just thinking out loud, no need to refactor it if it's not a clear win.

Reply:
My understanding of what's happening is trying to keep out any `WidgetState` logic from the painter?

Reply:
At this point the design of `ToggleablePainter` is looking dated. A painer should either take the resolved final color, or the unresolved WidgetState color; instead `ToggleablePainter` takes a half-resolved color with the old-style `xxxColor` API, which the caller needs to translate its `WidgetStateProperty` half-way to fit into. I agree that the current code might be the way to go, but I really think `ToggleablePainter` _should_ be made to take `WidgetState`s.